internal/locate/region_cache.go

```diff
@@ -2629,6 +2649,195 @@ func invokeKVStatusAPI(addr string, timeout time.Duration) (l livenessState) {
+		tsGradient = ss.tsCntSlidingWindow.Append(intervalAvgTimecost)
+	}
+	// Update avgScore & avgTimecost
+	avgScore := atomic.LoadUint64(&ss.avgScore)
+	if updGradient+0.1 <= float64(1e-9) && tsGradient-0.1 >= float64(1e-9) {
+		risenRatio := math.Min(float64(5.43), math.Abs(tsGradient/updGradient))
+		curAvgScore := math.Ceil(math.Min(float64(avgScore)*risenRatio+float64(1.0), float64(slowScoreMax)))
+		atomic.CompareAndSwapUint64(&ss.avgScore, avgScore, uint64(curAvgScore))
+	} else {
+		costScore := uint64(math.Ceil(math.Max(float64(slowScoreInitVal), math.Min(float64(2.71), 1.0+math.Abs(updGradient)))))
+		if avgScore <= slowScoreInitVal+costScore {
+			atomic.CompareAndSwapUint64(&ss.avgScore, avgScore, slowScoreInitVal)
+		} else {
+			atomic.CompareAndSwapUint64(&ss.avgScore, avgScore, avgScore-costScore)
+		}
+	}
+	atomic.CompareAndSwapUint64(&ss.avgTimecost, avgTimecost, ss.tsCntSlidingWindow.Avg())
+
+	// Resets the counter of inteval timecost
+	atomic.StoreUint64(&ss.intervalTimecost, 0)
+	atomic.StoreUint64(&ss.intervalUpdCount, 0)
+	return true
+}
+
+// recordSlowScoreStat records the timecost of each request.
+func (ss *SlowScoreStat) recordSlowScoreStat(timecost time.Duration) bool {
+	atomic.AddUint64(&ss.intervalUpdCount, 1)
+	avgTimecost := atomic.LoadUint64(&ss.avgTimecost)
+	if avgTimecost == 0 {
+		// Init the whole statistics with the original one.
+		atomic.StoreUint64(&ss.avgScore, slowScoreInitVal)
+		atomic.StoreUint64(&ss.avgTimecost, slowScoreInitTimeout)
+		atomic.StoreUint64(&ss.intervalTimecost, uint64(timecost/time.Microsecond))
+		return true
+	}
+	curTimecost := uint64(timecost / time.Microsecond)
+	if curTimecost >= slowScoreMaxTimeout {
+		// Current query is too slow to serve (>= 30s, max timeout of a request) in this tick.
+		atomic.StoreUint64(&ss.avgScore, slowScoreMax)
+		return false
+	}
+	atomic.AddUint64(&ss.intervalTimecost, curTimecost)
+	return true
+}
+
+func (ss *SlowScoreStat) markAlreadySlow() {
+	atomic.StoreUint64(&ss.avgScore, slowScoreMax)
+}
+
+func (ss *SlowScoreStat) isSlow() bool {
+	return ss.getSlowScore() >= slowScoreThreshold
+}
+
+// getSlowScore returns the slow score of store.
+func (s *Store) getSlowScore() uint64 {
+	return s.slowScore.getSlowScore()
+}
+
+// isSlow returns whether current Store is slow or not.
+func (s *Store) isSlow() bool {
+	return s.slowScore.isSlow()
+}
+
+// updateSlowScore updates the slow score of this store according to the timecost of current request.
+func (s *Store) updateSlowScoreStat() bool {
+	return s.slowScore.updateSlowScore()
+}
+
+// recordSlowScoreStat records timecost of each request.
+func (s *Store) recordSlowScoreStat(timecost time.Duration) bool {
+	return s.slowScore.recordSlowScoreStat(timecost)
+}
+
+func (s *Store) markAlreadySlow() {
+	s.slowScore.markAlreadySlow()
+}
+
+// asyncUpdateStoreSlowScore updates the slow score of each store periodically.
+func (c *RegionCache) asyncUpdateStoreSlowScore(interval time.Duration) {
```

Comment:
Ideally this asyncUpdateStoreSlowScore should not be called when the mode is not preferLeader. Or at least the expensive part of this function should not run.

Reply:
No recommend. 
`preferLeader` is a ReplicaReadMode for request-level or session-level setting. If and only if we have the global statistics of slowScores on all TiKV nodes by `asyncUpdateStoreSlowScore`, can we make the judgement whether a leader replica is slow or not and do following redirecting operations.
That is, we need the background statistics by the co-routine `asyncUpdateStoreSlowScore`.